keras_cv/training/contrastive/contrastive_trainer.py

```diff
@@ -0,0 +1,260 @@
+            )
+
+        self.augmenters = (
+            augmenter if type(augmenter) is tuple else (augmenter, augmenter)
+        )
+        self.encoder = encoder
+        self.projectors = (
+            projector if type(projector) is tuple else (projector, projector)
+        )
+
+        self.loss_metric = keras.metrics.Mean(name="loss")
+
+        self.include_probe = include_probe
+
+        if self.include_probe:
+            self.probing_top = layers.Dense(classes, name="linear_probe")
+            self.probe_loss_metric = keras.metrics.Mean(name="probe_loss")
+            self.probe_metrics = []
+
+    def compile(
+        self, probe_optimizer=None, probe_loss=None, probe_metrics=None, **kwargs
+    ):
+        super().compile(**kwargs)
+
+        if self.include_probe and not probe_optimizer:
+            raise ValueError(
+                "`probe_optimizer` must be specified when `include_probe` is `True`."
+            )
+
+        if self.include_probe and not probe_loss:
+            raise ValueError(
+                "`probe_loss` must be specified when `include_probe` is `True`."
+            )
+
+        if self.include_probe:
+            self.probe_loss = probe_loss
+            self.probe_optimizer = probe_optimizer
+            self.probe_metrics = probe_metrics or []
+
+    @property
+    def metrics(self):
+        metrics = [
+            self.loss_metric,
+        ]
+        if self.include_probe:
+            metrics += [self.probe_loss_metric]
+            metrics += self.probe_metrics
+        return super().metrics + metrics
+
+    def fit(
+        self,
+        x=None,
+        y=None,
+        sample_weight=None,
+        batch_size=None,
+        **kwargs,
+    ):
+        dataset = _convert_inputs_to_tf_dataset(
+            x=x, y=y, sample_weight=sample_weight, batch_size=batch_size
+        )
+
+        dataset = dataset.map(self.run_augmenters, num_parallel_calls=tf.data.AUTOTUNE)
+
+        return super().fit(x=dataset, **kwargs)
+
+    def run_augmenters(self, x, y=None):
+        if y is not None:
+            return (
+                x,
+                self.augmenters[0](x, training=True),
+                self.augmenters[1](x, training=True),
+                y,
+            )
+        else:
+            return (
+                x,
+                self.augmenters[0](x, training=True),
+                self.augmenters[1](x, training=True),
+            )
+
+    def train_step(self, data):
+        if self.include_probe:
+            if len(data) != 4:
+                raise ValueError(
+                    "Targets must be provided when `include_probe` is True"
+                )
+            images, augmented_images_0, augmented_images_1, labels = data
+        else:
+            if len(data) != 3:
+                raise ValueError(
+                    "Targets must not be provided when `include_probe` is False"
```

Comment:
you can just add `None` on to data if its len is 3, the if the labels should be provided (i.e. in the probe loop) then you raise an error if labels are None.'

This probably belongs in a helper method regardless (`self.parse_inputs(data)` perhaps?)

Reply:
I think with the restructuring of the data as a dictionary this is pretty clean as a guard clause in `fit`